segmentation/utils/spatial_analysis.py

```diff
@@ -0,0 +1,245 @@
+        marker_num: number of markers in expresion data
+        dist_lim: threshold for spatial enrichment distance proximity
+        cell_label_idx: column with cell labels
+        bootstrap_num: number of permutations
+
+    Returns
+        close_num_rand: random positive marker counts
+            for every permutation in the bootstrap"""
+
+    # Create close_num_rand
+    close_num_rand = np.zeros((
+        marker_num, marker_num, bootstrap_num), dtype='int')
+
+    for j in range(0, marker_num):
+        for k in range(0, marker_num):
+            for r in range(0, bootstrap_num):
+                # Select same amount of random cell labels as positive ones in same marker in close_num
+                marker1labelsrand = patient_data[
+                    patient_data.columns[cell_label_idx]].sample(
+                        n=marker1_num[j], replace=True)
+                marker2labelsrand = patient_data[
+                    patient_data.columns[cell_label_idx]].sample(
+                        n=marker2_num[k], replace=True)
+                # Subset the distance matrix to only include positive randomly selected cell labels
+                rand_trunc_dist_mat = dist_mat[np.ix_(
+                    np.asarray(marker1labelsrand - 1), np.asarray(
+                        marker2labelsrand - 1))]
+                # Binarize the truncated distance matrix to only include cells within distance limit
+                rand_trunc_dist_mat_bin = np.zeros(rand_trunc_dist_mat.shape, dtype='int')
+                rand_trunc_dist_mat_bin[rand_trunc_dist_mat < dist_lim] = 1
+                # Record the number of interactions and store in close_num_rand in the index
+                # corresponding to both markers, for every permutation
+                close_num_rand[j, k, r] = np.sum(np.sum(rand_trunc_dist_mat_bin))
+    return close_num_rand
+
+
+def calculate_enrichment_stats(close_num, close_num_rand):
+    """Calculates z score and p values from spatial enrichment analysis.
+
+    Args:
+        close_num: marker x marker matrix with counts for cells
+            positive for corresponding markers
+        close_num_rand: random positive marker counts
+            for every permutation in the bootstrap
+
+    Returns:
+        z: z scores for corresponding markers
+        muhat: predicted mean values of close_num_rand random distribution
+        sigmahat: predicted standard deviation values of close_num_rand
+            random distribution
+        p: p values for corresponding markers, for both positive
+            and negative enrichment
+        h: matrix indicating whether
+            corresponding marker interactions are significant
+        adj_p: fdh_br adjusted p values"""
+    # Get the number of markers and number of permutations
+    marker_num = close_num.shape[0]
+    bootstrap_num = close_num_rand.shape[2]
+
+    # Create z, muhat, sigmahat, and p
+    z = np.zeros((marker_num, marker_num))
```

Comment:
lets instead create a single array to hold all of these values. That way, we can save this single array at the end. 

The dimensions should be [stats x marker1 x marker2]. Let's make a separate p_pos and p_neg array, that way everything will have the same shape. 

Then, we can use xarray to label the dimensions. Take a look at https://github.com/angelolab/segmentation/blob/76fbf4682f108bd047f878262337c0be9a3c26e5/segmentation/utils/data_utils.py#L48 to see how to create an xarray